File: UnitTest/Zeze/Trans/TestCheckpointModeTable.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Zeze.Transaction;
using Zeze.Util;

namespace UnitTest.Zeze.Trans
{
    [TestClass]
    public class TestCheckpointModeTable
    {
        [TestInitialize]
        public void TestInit()
        {
            demo.App.Instance.Start();
        }

        [TestCleanup]
        public void TestCleanup()
        {
            demo.App.Instance.Stop();
        }

        private async Task Check(int expect)
        {
            Assert.IsTrue(ResultCode.Success == await demo.App.Instance.Zeze.NewProcedure(
                async () =>
                {
                    var value = await demo.App.Instance.demo_Module1.TableImportant.GetOrAddAsync(1);
                    return value.Int1 == expect ? ResultCode.Success : ResultCode.LogicError;
                },
                "TestCheckpointModeTable.Check").CallAsync());
        }

        [TestMethod]
        public async Task Test1()
        {
            Assert.IsTrue(ResultCode.Success == await demo.App.Instance.Zeze.NewProcedure(
                async () =>
                {
                    var value = await demo.App.Instance.demo_Module1.TableImportant.GetOrAddAsync(1);
                    value.Int1 = 0;
                    return ResultCode.Success;
                },
                "TestCheckpointModeTable.Init").CallAsync());
            await Check(0);

            int sum = 0;
            {
                Task[] tasks = new Task[1000];
                for (int i = 0; i < tasks.Length; ++i)
                {
                    tasks[i] = demo.App.Instance.Zeze.NewProcedure(Add, "TestCheckpointModeTable.Add").CallAsync();
                }
                await Task.WhenAll(tasks);
                sum += tasks.Length;
                await Check(sum);
            }

            {
                Task[] tasks = new Task[1000];
                for (int i = 0; i < tasks.Length; ++i)
                {
                    tasks[i] = demo.App.Instance.Zeze.NewProcedure(Add2, "TestCheckpointModeTable.Add2").CallAsync();
                }
                await Task.WhenAll(tasks);
                sum += tasks.Length;
                await Check(sum);
            }
        }

        private async Task<long> Add()
        {
            var value = await demo.App.Instance.demo_Module1.TableImportant.GetOrAddAsync(1);
            value.Int1++;
            return ResultCode.Success;
        }

        private async Task<long> Add2()
        {
            var value = await demo.App.Instance.demo_Module1.TableImportant.GetOrAddAsync(1);
            value.Int1++;
            var value2 = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(1);
            value2.Int1++;
            return ResultCode.Success;
        }
    }
}

File: ZezeJava/ZezeJavaTest/src/UnitTest/Zeze/Trans/TestCheckpointModeTable.java
package UnitTest.Zeze.Trans;

import java.util.concurrent.Future;
import Zeze.Transaction.DispatchMode;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import Zeze.Transaction.Procedure;

public class TestCheckpointModeTable{

	@Before
	public final void testInit() throws Exception {
		demo.App.getInstance().Start();
	}

	@After
	public final void testCleanup() throws Exception {
		demo.App.getInstance().Stop();
	}

	private static void Check(int expect) throws Exception {
		Assert.assertEquals(Procedure.Success, demo.App.getInstance().Zeze.newProcedure(() -> {
					var value = demo.App.getInstance().demo_Module1.getTableImportant().getOrAdd(1L);
					return value.getInt1() == expect ? Procedure.Success : Procedure.LogicError;
		}, "TestCheckpointModeTable.Check").call());
	}

	@Test
	public final void test1() throws Exception {
		Assert.assertEquals(Procedure.Success, demo.App.getInstance().Zeze.newProcedure(() -> {
					var value = demo.App.getInstance().demo_Module1.getTableImportant().getOrAdd(1L);
					value.setInt1(0);
					return Procedure.Success;
		}, "TestCheckpointModeTable.Init").call());
		Check(0);

		int sum = 0; {
			Future<?>[] tasks = new Future[1000];
			for (int i = 0; i < tasks.length; ++i) {
				tasks[i] = Zeze.Util.Task.runUnsafe(
						demo.App.getInstance().Zeze.newProcedure(TestCheckpointModeTable::Add, "TestCheckpointModeTable.Add"),
						null, null, DispatchMode.Normal);
			}
			for (Future<?> task : tasks) {
				try {
					task.get();
				} catch (Throwable e) {
					// print stacktrace.
					e.printStackTrace();
				}
			}
			sum += tasks.length;
			Check(sum);
		}

		{
			Future<?>[] tasks = new Future[1000];
			for (int i = 0; i < tasks.length; ++i) {
				tasks[i] = Zeze.Util.Task.runUnsafe(
						demo.App.getInstance().Zeze.newProcedure(TestCheckpointModeTable::Add2, "TestCheckpointModeTable.Add2"),
						null, null, DispatchMode.Normal);
			}
			for (Future<?> task : tasks) {
				try {
					task.get();
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
			sum += tasks.length;
			Check(sum);
		}
	}

	private static long Add() {
		var value = demo.App.getInstance().demo_Module1.getTableImportant().getOrAdd(1L);
		value.setInt1 (value.getInt1() + 1);
		return Procedure.Success;
	}

	private static long Add2() {
		var value = demo.App.getInstance().demo_Module1.getTableImportant().getOrAdd(1L);
		value.setInt1(value.getInt1() + 1);
		var value2 = demo.App.getInstance().demo_Module1.getTable1().getOrAdd(1L);
		value2.setInt1 ( value2.getInt1() + 1);
		return Procedure.Success;
	}
}
